content/blog/2019-12-18-dependently-typed-language/index.md

```diff
@@ -0,0 +1,206 @@
+testing to see the reduced value and type of various terms.
+
+## Evaluation
+For the evaluation, I decided to see what kinds of theorems I could prove from the Coq standard library.
+I decided to choose theorems from the [Logic](https://coq.inria.fr/library/Coq.Init.Logic.html) library, which
+seemed like one of the simplest libraries. This had definitions for things like True and False (or Top and
+Bottom respectively), propositional or and and, if and only if, etc. I was able to state and prove a few
+theorems.
+
+First we can make some definitions that the library does:
+```
+let False = [a:*]a in
+let True = [a: *][I: a]a in
+let not = [A: *][_: A] False in
+let sum = (\A: *)(\B: *) [R: *][f: [_: A]R][g: [_: B]R]R in
+let or_introl = (\A: *)(\B: *)(\x: A)(\R: *)(\f: [_: A]R)(\g: [_: B]R)(f x) in
+let or_intror = (\A: *)(\B: *)(\y: B)(\R: *)(\f: [_: A]R)(\g: [_: B]R)(g y) in
+...
+```
+These are the same definitions as in Coq, with slightly less elegant syntax, especially because
+inductive types have to be encoded.
+
+Now here are some proofs about `and`:
+```
+let prod = (\A: *)(\B: *) [R: *] [_: [_: A][_: B]R] R in
+let conj = (\A: *)(\B: *)(\x: A)(\y: B)(\R: *)(\f: [_: A][_: B]R)((f x) y) in
+
+Theorem proj1 = [A: *][B: *][_:  ((prod A) B)]A with
+    Proof (\A: *)(\B: *)(\p: ((prod A) B))((p A) (\x: A)(\y: B)x);
+
+Theorem proj2 = [A: *][B: *][_:  ((prod A) B)]B with
+    Proof (\A: *)(\B: *)(\p: ((prod A) B))((p B) (\x: A)(\y: B)y);
+```
+Here, `prod` represents propositional `and`, while `conj` is the constructor which produces a term
+of that type. Then, Theorem `proj1` and `proj2` show that $\forall A, B. A \wedge B \implies A$ and
+$\forall A, B. A \wedge B \implies B$. When we consider these terms as encodings of pairs, `proj1`
+and `proj2` are simply the `fst` and `snd` functions like in OCaml.
+
+Here are some more proofs relating to "if and only if":
+```
+let iff = (\X: *)(\Y: *) ((prod [_: X]Y) [_: Y]X) in
+
+Theorem iff_refl = [A: *]((iff A) A) with
+Proof (\A: *)((((conj [_: A]A) [_: A]A) (\x:A)x) (\x:A)x);
+
+let swap = (\A1:*)(\B1:*)(\p1: ((prod A1) B1))
+    ( ( ((conj B1) A1)
+        (((proj2 A1) B1) p1)
+      )
+      (((proj1 A1) B1) p1)
+    ) in
+
+Theorem iff_sym = [A2: *][B2: *][_: ((iff A2) B2)]((iff B2) A2) with
+Proof (\A2:*)(\B2:*)(\A_imp_B2: ((iff A2) B2))
+    (((swap [_: A2]B2) [_: B2]A2) A_imp_B2);
+```
+
+`iff_refl` claims that if and only if is reflexive, while `iff_sym` claims that it is symmetric.
+`iff_sym` was actually a case where bugs in substitution came up, so I just used different names for
+abstractions/products.
+
+Interestingly, the last two theorems actually took about a second to typecheck, which is quite
+surprising because Coq can typecheck these terms extremely quickly. This may be because Coq is based on
+the Calculus of *Inductive* Constructions, which is quite different from CoC because it introduces
+universes. This increased expressivity probably allows for faster implementations.
```

Comment:
Yeah I wish I had some more time to do a rigorous profiling. Just now I tried a very simple profiling by writing code inside the function to get a total of how long it ran, but because there are lots of recursive calls it seemed like it was double counting a lot so I couldn't get a good estimate of where it was spending a lot of time...